core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+									// Redirect to CID fixed inside of toSubdomainURL()
+									http.Redirect(w, r, newURL, http.StatusMovedPermanently)
+									return
+								}
+							}
+						}
+
+						// Rewrite the path to not use subdomains
+						r.URL.Path = pathPrefix + r.URL.Path
+						// Serve path request
+						childMux.ServeHTTP(w, r)
+						return
+					}
+					// If not, resource does not exist on the subdomain gateway, return 404
+					http.NotFound(w, r)
+					return
+				}
+			}
+			// We don't have a known gateway. Fallback on DNSLink lookup
+
+			// Wildcard HTTP Host check:
+			// 1. is wildcard DNSLink enabled (Gateway.NoDNSLink=false)?
+			// 2. does Host header include a fully qualified domain name (FQDN)?
+			// 3. does DNSLink record exist in DNS?
+			if !cfg.Gateway.NoDNSLink && isDNSLinkRequest(r, n) {
+				// rewrite path and handle as DNSLink
+				r.URL.Path = "/ipns/" + stripPort(r.Host) + r.URL.Path
+				childMux.ServeHTTP(w, r)
+				return
+			}
+
+			// else, treat it as an old school gateway, I guess.
+			childMux.ServeHTTP(w, r)
+		})
+		return childMux, nil
+	}
+}
+
+// isDNSLinkRequest returns bool that indicates if request
+// should return data from content path listed in DNSLink record (if exists)
+func isDNSLinkRequest(r *http.Request, n *core.IpfsNode) bool {
+	ctx, cancel := context.WithCancel(n.Context())
+	defer cancel()
+
+	fqdn := stripPort(r.Host)
+	if len(fqdn) > 0 && isd.IsDomain(fqdn) {
+		// Confirm DNSLink was not disabled for the fqdn or globally
+		name := "/ipns/" + fqdn
+		// check if DNSLink exists
+		_, err := n.Namesys.Resolve(ctx, name, nsopts.Depth(1))
+		if err == nil || err == namesys.ErrResolveRecursion {
+			return true
+		}
+	}
+	return false
+}
+
+func isSubdomainNamespace(ns string) bool {
+	switch ns {
+	case "ipfs", "ipns", "p2p", "ipld", "api":
+		return true
+	default:
+		return false
+	}
+}
+
+// Parses Host header of a subdomain-based URL and returns it's components
+// Note: hostname is host + optional port
+func parseSubdomains(hostHeader string) (hostname, ns, rootID string, ok bool) {
+	parts := subdomainGatewayRegex.FindStringSubmatch(hostHeader)
+	if len(parts) < 4 || !isSubdomainNamespace(parts[2]) {
+		return "", "", "", false
+	}
+	return parts[3], parts[2], parts[1], true
+}
+
+// Converts a hostname/path to a subdomain-based URL, if applicable.
+func toSubdomainURL(hostname, path string, query string) (url string, ok bool) {
+	parts := strings.SplitN(path, "/", 4)
+
+	var ns, rootID, rest string
+	switch len(parts) {
+	case 4:
+		rest = parts[3]
+		fallthrough
+	case 3:
+		ns = parts[1]
+		rootID = parts[2]
+	default:
+		return "", false
+	}
+
+	if !isSubdomainNamespace(ns) {
+		return "", false
+	}
+
+	// add prefix if query is present
+	if query != "" {
+		query = "?" + query
+	}
+
+	if ns == "api" || ns == "p2p" {
+		// API and P2P proxy use the same paths on subdomains:
+		// api.hostname/api/.. and p2p.hostname/p2p/..
+		return fmt.Sprintf(
```

Comment:
Are we going to get any double slashes here? We might want to use URL from `net/url` to format the URL.

Reply:
Done in 53413cefe